src/libs/sfdk/asynchronous.h

```diff
@@ -0,0 +1,57 @@
+/****************************************************************************
+**
+** Copyright (C) 2019 Open Mobile Platform LLC.
+** Contact: http://jolla.com/
+**
+** This file is part of Qt Creator.
+**
+** GNU Lesser General Public License Usage
+** Alternatively, this file may be used under the terms of the GNU Lesser
+** General Public License version 2.1 as published by the Free Software
+** Foundation and appearing in the file LICENSE.LGPL included in the
+** packaging of this file.  Please review the following information to
+** ensure the GNU Lesser General Public License version 2.1 requirements
+** will be met: http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.
+**
+** Other Usage
+**
+** Alternatively, this file may be used in accordance with the terms and
+** conditions contained in a signed written agreement between you and Digia.
+**
+****************************************************************************/
+
+#pragma once
+
+#include "sfdkglobal.h"
+
+#include <QEventLoop>
+#include <QPointer>
+
+#include <tuple>
+#include <utility>
+
+namespace Sfdk {
+
+template<typename... Ts>
+using Functor = std::function<void(Ts...)>;
+
+template<typename Fn, typename... Args>
+void callIf(const QPointer<const QObject> &context, const Fn &fn, Args&&... args)
+{
+    if (context)
+        fn(std::forward<Args>(args)...);
+}
+
+template<typename Fn, typename... InArgs, typename... OutArgs>
+void execAsynchronous(std::tuple<OutArgs&...> outArgs, const Fn &fn, InArgs&&... inArgs)
+{
+	QEventLoop loop;
+	auto whenDone = [&loop, &outArgs](OutArgs...as) {
+		loop.quit();
+		outArgs = std::tie(as...);
+	};
+	fn(std::forward<InArgs>(inArgs)..., &loop, whenDone);
+	loop.exec();
```

Comment:
If fn() happens to be synchronous (i.e. calls whenDone before returning), loop.exec() blocks forever. I would consider setting a flag in whenDone, and not calling exec() if whenDone has been already called. This way the caller would not have to care whether the implementation is asynchronous or not.

Reply:
It is designed so that it is guaranteed that any two asynchronous calls would complete in the same order in which they were initiated. It makes things easier a lot on the caller side. For that to work, every asynchronous function must be really asynchronous. If you need to fail an asynchronous function immediately, use something like `commandQueue()->enqueueCheckPoint(context, std::bind(functor, false))` to achieve proper serialization.

Reply:
There was no mention about the order preservation, so I didn't know that. This could of course be implemented so that it would do also that even with synchronous functions, but that would require a bit more code here.

My issue is not with failing function immediately, more that when the function implementation is more or less no-op, making it asynchronous feels like a lot of wasted lines of code.